app/HutchAgent/Services/OpenIdIdentityService.cs

```diff
@@ -178,33 +178,34 @@ public async Task<string> RequestClientAccessToken(string clientId, string secre
     // return the tokens for use
     return tokenResponse.AccessToken;
   }
-  
+
   /// <summary>
   /// Refreshing access token
   /// </summary>
   /// <param name="options"></param>
   /// <param name="refreshToken"></param>
   /// <returns></returns>
   public async Task<(string access, string refresh)> RefreshAccessToken(OpenIdOptions options, string refreshToken) =>
-    await RefreshAccessToken(options.ClientId, refreshToken);
-  
+    await RefreshAccessToken(options.ClientId, options.ClientSecret, refreshToken);
+
   /// <summary>
   /// Refreshing token using refresh token endpoint
   /// </summary>
   /// <param name="clientId"></param>
+  /// <param name="secret"></param>
   /// <param name="refreshToken"></param>
   /// <returns></returns>
   /// <exception cref="InvalidOperationException"></exception>
-  public async Task<(string access, string refresh)> RefreshAccessToken(string clientId,
+  public async Task<(string access, string refresh)> RefreshAccessToken(string clientId, string secret,
     string refreshToken)
   {
     var disco = await GetDiscoveryDocument();
 
-    // Make a password token request for a user
     var tokenResponse = await _http.RequestRefreshTokenAsync(new()
     {
       Address = disco.TokenEndpoint,
       ClientId = clientId,
+      ClientSecret = secret,
```

Comment:
I think, however it should be optional; certainly in keycloak you can configure a client as not requiring authentication, and therefore a secret wouldn't be expected.

So in Hutch terms, I would suggest if secret was null or whitespace only, then don't send it?

Reply:
(theoretically the original non-refresh token code i wrote should behave that way already? but i can't remember if it actually does)